server/src/main/java/org/apache/kafka/server/ClientMetricsManager.java

```diff
@@ -459,6 +469,48 @@ Timer expirationTimer() {
         return expirationTimer;
     }
 
+    // Visible for testing
+    Map<String, Uuid> clientConnectionIdMap() {
+        return clientConnectionIdMap;
+    }
+
+    private final class ClientConnectionDisconnectListener implements ConnectionDisconnectListener {
+
+        private final Cache<Uuid, ClientMetricsInstance> clientInstanceCache;
+        private final Map<String, Uuid> clientConnectionIdMap;
+        private final ClientMetricsStats clientMetricsStats;
+
+        ClientConnectionDisconnectListener(
+            Cache<Uuid, ClientMetricsInstance> clientInstanceCache,
+            Map<String, Uuid> clientConnectionIdMap,
+            ClientMetricsStats clientMetricsStats
+        ) {
+            this.clientInstanceCache = clientInstanceCache;
+            this.clientConnectionIdMap = clientConnectionIdMap;
+            this.clientMetricsStats = clientMetricsStats;
+        }
+
+        @Override
+        public void onDisconnect(String connectionId) {
+            log.trace("Removing client connection id [{}] from the client instance cache", connectionId);
+
+            Uuid clientInstanceId = clientConnectionIdMap.remove(connectionId);
+            if (clientInstanceId == null) {
+              log.trace("Client connection id [{}] is not found in the client instance cache", connectionId);
+              return;
+            }
+
+            // Unregister the client instance metrics from the broker metrics.
+            clientMetricsStats.unregisterClientInstanceMetrics(clientInstanceId);
+
+            ClientMetricsInstance clientInstance = clientInstanceCache.get(clientInstanceId);
+            if (clientInstance != null) {
+                clientInstance.cancelExpirationTimerTask();
+                clientInstanceCache.remove(clientInstanceId);
```

Comment:
This means that every time we restart a broker, we could prematurely remove a live clientInstanceId. Do we gain more than we lose?

Reply:
If a broker is restarted then anyways the cache will be clean. I am not sure if I understood the question right. I must be missing somehting to understand.

Reply:
You are right. Restarting a broker doesn't cause additional issues.

> Though terminating is good enough to not accept further push telemetry calls from same client but if we remove the instance from cache on terminating flag then it will open us for attacks with rogue clients which can send telemetry push request with terminating flag as true and we will skip push time based throttling for this special push request. Hence it would be better to keep these 2 semantics separately.

Does that mean we shouldn't evict from the cache for clients marked as terminating? If we do that, do we really get any benefit since most of the clients will go through the terminating process?

Reply:
>Does that mean we shouldn't evict from the cache for clients marked as terminating? If we do that, do we really get any benefit since most of the clients will go through the terminating process?

What I meant was that we should keep terminating flag behaviour separately from cache eviction. We shouldn't evict from cache based on terminating flag but should only evict when connection is dropped.
Say, there is a rogue client which sends telemetry push requests always with terminating as true and on a same open connection. Then because the instance will be in cache as connection is still open then we can deny the request. Hence to protect broker and down stream systems, we should always act on client disconnection.

Reply:
So, we will always evict the client from the cache on connection close. But if a client is terminating, the broker will still remember the client uuid somewhere else to protect against rogue client. Is that correct?

Reply:
Yes we will evict the instance on connection close. But once the instance is evicted then the broker can't remember the UUID. The 2 scenarios I am talking about are

1. On existing connection a rogue client sends all telemetry push metrics with flag as terminating: true: a) As the connection is not closed hence instance is not evicted, though client sends the rpc as terminating: true but we have the instance cached which errors out the requests.
2. When connection is closed: The scenario is same as earlier, before this PR. When a new client is spawned after closed connection then anyways new instance id is generated which has it's own respective checks.

By not evicting client instances on terminating:true we are handling the 1st scenario mentioned above.